src/protocol/boolean/comparison.rs

```diff
@@ -2,10 +2,126 @@ use super::or::or;
+// <https://doi.org/10.1007/978-3-540-71677-8_23>
+//
+// The version in the paper tests c_1 < x < c_2. For us, c_1 is zero (which eliminates the `c < c_1`
+// case enumerated in the paper), we test ≤ rather than <, and we finally return `x > c_2` which is
+// ~(0 ≤ x ≤ c_2).
+//
+// The remainder of this description names the variables consistently with other routines in this
+// file, which map to the paper as follows:
+//
+// Paper  Ours
+// x      a
+// c      b
+// c_1    0
+// c_2    c
+//
+// Goal: Compute `a > c` as `~(0 ≤ a ≤ c)`
+//
+// Strategy:
+//  1. Generate random r
+//  2. Reveal b = a + r
+//  3. Derive bounds r_low and r_high from public values, such that the desired result is a simple
+//     function of `r_low < r` and `r < r_high`.
+//
+// Case 1: b > c
+// b - c - 1 < r < b + 1
+//  ⟺  b - c ≤ r ≤ b
+//  ⟺  b - b ≤ b - r ≤ b - (b - c)
+//  ⟺  0 ≤ a ≤ c
+//  ⟺  ~(a > c)
+//
+// Case 2: b ≤ c
+//  b < r < b + p - c
+//  ⟺  b - (b + p - c) < b - r < b - b
+//  ⟺  -(p - c) < a < 0
+//  ⟺  a > c
+//
+/// # Errors
+/// Lots of things may go wrong here, from timeouts to bad output. They will be signalled
+/// back via the error response
+pub async fn greater_than_constant<F, C, S>(
+    ctx: C,
+    record_id: RecordId,
+    rbg: &RandomBitsGenerator<F, S, C>,
+    a: &S,
+    c: u128,
+) -> Result<S, Error>
+where
+    F: Field,
+    C: Context + RandomBits<F, Share = S>,
+    S: ArithmeticSecretSharing<F> + BasicProtocols<C, F>,
+{
+    use GreaterThanConstantStep as Step;
+
+    let r = rbg.generate().await?;
+
+    let b = (r.b_p.clone() + a)
+        .reveal(ctx.narrow(&Step::Reveal), record_id)
+        .await?;
+
+    let r_lo;
+    let r_hi;
+    let invert;
+    if b.as_u128() > c {
+        r_lo = b.as_u128() - c - 1;
+        r_hi = b.as_u128() + 1;
+        invert = true;
+    } else {
+        r_lo = b.as_u128();
+        r_hi = F::PRIME.into() + b.as_u128() - c;
+        invert = false;
+    }
+
+    let r_gt_r_lo =
+        bitwise_greater_than_constant(ctx.narrow(&Step::CompareLo), record_id, &r.b_b, r_lo)
+            .await?;
+    let r_lt_r_hi =
+        bitwise_less_than_constant(ctx.narrow(&Step::CompareHi), record_id, &r.b_b, r_hi).await?;
+
+    let result = r_gt_r_lo
+        .multiply(&r_lt_r_hi, ctx.narrow(&Step::Multiply), record_id)
+        .await?;
+
+    if invert {
+        Ok(S::share_known_value(&ctx, F::ONE) - &result)
+    } else {
+        Ok(result)
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum GreaterThanConstantStep {
+    Reveal,
+    CompareLo,
+    CompareHi,
+    Multiply,
+}
+
+impl crate::protocol::Substep for GreaterThanConstantStep {}
+
+impl AsRef<str> for GreaterThanConstantStep {
+    fn as_ref(&self) -> &str {
+        match self {
+            Self::Reveal => "reveal",
+            Self::CompareLo => "compare_lo",
+            Self::CompareHi => "compare_hi",
+            Self::Multiply => "multiply",
```

Comment:
```suggestion
            Self::And => "and",
```

It's a logical AND, so maybe this is clearer.